File: src/apps/EventViewer/src/DevicesController.m
#import "DevicesController.h"
#import "libkrbn/libkrbn.h"
#import <pqrs/weakify.h>

@interface DevicesController ()

@property(unsafe_unretained) IBOutlet NSTextView* textView;

- (void)callback:(NSString*)filePath;

@end

static void staticCallback(const char* filePath,
                           void* context) {
  DevicesController* p = (__bridge DevicesController*)(context);
  if (p && filePath) {
    [p callback:[NSString stringWithUTF8String:filePath]];
  }
}

@implementation DevicesController

- (void)setup {
  libkrbn_enable_device_details_json_file_monitor(staticCallback,
                                                  (__bridge void*)(self));
}

- (void)dealloc {
  libkrbn_disable_device_details_json_file_monitor();
}

- (void)callback:(NSString*)filePath {
  @weakify(self);
  dispatch_async(dispatch_get_main_queue(), ^{
    @strongify(self);
    if (!self) {
      return;
    }

    NSString* string = [NSString stringWithContentsOfFile:filePath
                                                 encoding:NSUTF8StringEncoding
                                                    error:nil];
    if (string) {
      NSTextStorage* textStorage = self.textView.textStorage;
      NSFont* font = [NSFont fontWithName:@"Menlo" size:11];

      [textStorage beginEditing];
      [textStorage setAttributedString:[[NSAttributedString alloc] initWithString:string
                                                                       attributes:@{
                                                                         NSForegroundColorAttributeName : NSColor.textColor,
                                                                         NSFontAttributeName : font,
                                                                       }]];
      [textStorage endEditing];
    }
  });
}

@end

File: src/apps/EventViewer/src/DevicesController.swift
import Cocoa

private func callback(_ filePath: UnsafePointer<Int8>?, _ context: UnsafeMutableRawPointer?) {
  if filePath == nil { return }

  let path = String(cString: filePath!)
  let obj: DevicesController? = unsafeBitCast(context, to: DevicesController.self)

  guard let message = try? String(contentsOfFile: path, encoding: .utf8) else { return }

  DispatchQueue.main.async { [weak obj] in
    guard let obj = obj else { return }

    guard let textStorage = obj.textView!.textStorage else { return }
    guard let font = NSFont(name: "Menlo", size: 11) else { return }
    let attributedString = NSAttributedString(string: message, attributes: [
      NSAttributedString.Key.font: font,
      NSAttributedString.Key.foregroundColor: NSColor.textColor,
    ])

    textStorage.beginEditing()
    textStorage.setAttributedString(attributedString)
    textStorage.endEditing()
  }
}

@objc
public class DevicesController: NSObject {
  @IBOutlet var textView: NSTextView?

  deinit {
    libkrbn_disable_device_details_json_file_monitor()
  }

  @objc
  public func setup() {
    let obj = unsafeBitCast(self, to: UnsafeMutableRawPointer.self)
    libkrbn_enable_device_details_json_file_monitor(callback, obj)
  }
}
